packages/core/artifact-testing/src/experimental.test.ts

```diff
@@ -2,12 +2,188 @@
+          instructions: 'Research information and entities related to the selected objects.',
+          tools: ['search/web_search'],
+        },
+        {
+          instructions:
+            'Based on your research find matching entires that are already in the graph. Do exaustive research.',
+          tools: ['search/local_search'],
+        },
+        {
+          instructions: 'Add researched data to the graph. Make connections to existing objects.',
+          tools: ['search/local_search', 'graph/writer'],
+        },
+      ],
+    };
+
+    const blueprint = BlueprintParser.create().parse(blueprintDefinition);
+
+    const researchQueue = queues.create();
+    const tools = new ToolRegistry(
+      [
+        createExaTool({ apiKey: EXA_API_KEY }),
+        createLocalSearchTool(db, researchQueue),
+        createGraphWriterTool({
+          db,
+          queue: researchQueue,
+          schema: DataTypes,
+          onDone: async (objects) => {
+            await researchQueue.append(objects);
+          },
+        }),
+      ].filter(isNonNullable),
+    );
+
+    const company = db.add(
+      Obj.make(DataType.Organization, {
+        name: 'Notion',
+        website: 'https://www.notion.com',
+      }),
+    );
+    await db.flush({ indexes: true });
+
+    const machine = new BlueprintMachine(tools, blueprint);
+    const { client } = serviceContainer.getService(AiService);
+    setConsolePrinter(machine, true);
+    await machine.runToCompletion({ aiClient: client, input: [company] });
+
+    log.info('researched', {
+      objects: await researchQueue.queryObjects(),
+    });
+  });
+
+  test('circuit', { timeout: 120_000 }, async () => {
+    // TODO(dmaretskyi): Can we type graph inputs/outputs?
+    // TODO(dmaretskyi): Store in ECHO.
+    const gptCircuit = () => {
+      const model = ComputeGraphModel.create();
+      model.builder
+        .createNode({ id: 'gpt1-INPUT', type: NODE_INPUT })
+        .createNode({ id: 'gpt1-GPT', type: 'gpt' })
+        .createNode({ id: 'gpt1-OUTPUT', type: NODE_OUTPUT })
+        .createEdge({ node: 'gpt1-INPUT', property: 'prompt' }, { node: 'gpt1-GPT', property: 'prompt' })
+        .createEdge({ node: 'gpt1-GPT', property: 'text' }, { node: 'gpt1-OUTPUT', property: 'text' });
+
+      return model;
+    };
+
+    const runtime = new TestRuntime(serviceContainer);
+    runtime.registerGraph('dxn:compute:gpt1', gptCircuit());
+
+    const { text } = await runtime
+      .runGraph<GptOutput>('dxn:compute:gpt1', ValueBag.make({ prompt: 'What is the meaning of life?' }))
+      .pipe(Effect.flatMap(ValueBag.unwrap), Effect.scoped, Effect.runPromise);
+
+    log.info('text', { text });
+  });
+
+  test('conversation', { timeout: 120_000 }, async () => {
+    const { client } = serviceContainer.getService(AiService);
+    const session = new AISession({
+      operationModel: 'configured',
+    });
+
+    // TODO(dmaretskyi): Use tool registry.
+    const sage = createTool('test', {
```

Comment:
funny but let's start to develop something mildly more useful in tests like the calculator in my ai.test.ts since it actually uses params and schema. building up a repertoire of tools via tests would also be useful.